### PR TITLE
First snap flow for java

### PR DESCRIPTION
**Pull request by igorljubuncic**
Added Java FSF using the existing template. I did not edit the snapcraft.yaml content or make any changes to the language-common sections.

### QA

- DEMO: https://snapcraft-io-canonical-websites-pr-1558.run.demo.haus/
- open Java section of first snap flow on home page
- go through the steps

screenshot 2019-02-05 at 12 20 45

### CODE REVIEW

**Comment by bartaz** [comment]
Code looks good, thanks.

@evandandrea Does the content here need review/QA? Or is it good to be merged already?

**Comment by igorljubuncic** [comment]
At this point, I have only made the Java FSF consistent with the expected language guide flow. The functional and logic changes is step 2, which we agreed to address once all guides use the same flow, and then we can make changes across all languages.

I will look at the styling suggestions and implement those and mark the unresolved functional/logic ones - as they may be common across other guides.

**Comment by evandandrea** [comment]
Confirmed the snap builds when dropping `base: core18`, but I am unable to test it.

**Comment by igorljubuncic** [comment]
Accepted all changes and added a new commit.

**Comment by igorljubuncic** [comment]
Base heading still present as it is hard-coded in templates/first-snap/package.html.

File: webapp/first_snap/content/java/package.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  description: |
+    Freeplane is a free and open source software application that supports thinking, sharing information and getting things done at work, in school and at home. The core of the software is tools for mind mapping (also known as concept mapping or information mapping) and using mapped information.
+explain:
+  metadata:
+    - text: The name must be unique in the Snap Store. Valid snap names consist of lower-case alphanumeric characters and hyphens. They cannot be all numbers. They also cannot start or end with a hyphen.
+    - code: |
+        name: freeplane
+    - warning: |
+        <span class="p-notification__status">Snap names are globally unique.</span> You should change any instance of <code>freeplane</code> to <code>freeplane-{name}</code> (where <code>{name}</code> is your name) before continuing to avoid conflicting with the official freeplane snap.
+    - text: Versions carry no semantic meaning in snaps.
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: A bit terse. Maybe, "Versions carry no semantic meaning in snaps. Use text that best identifies that release, it won't be used to compare against the preceding or following one."

File: webapp/first_snap/content/java/package.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    - code: |
+        description: |
+  security:
+    - text: The next section describes the level of confinement applied to your app.
+    - code: |
+        confinement: devmode
+    - text: Snaps are containerised to ensure more predictable application behaviour and greater security.
+    - text: It’s best to start a snap with the confinement in warning mode, rather than strictly applied. This is indicated through the <code>devmode</code> keyword.
+    - text: Once an app is working well in devmode, you can review confinement violations, add appropriate interfaces, and switch to strict confinement.
+  parts:
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: Congratulations, you found a bug:
screenshot 2019-02-07 at 18 30 01

We should fix as part of this PR.

File: webapp/first_snap/content/java/package.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    - text: |
+        Parts define how to build your app. Parts can be anything: programs, libraries, or other assets needed to create and run your application. Parts canpoint to local directories, remote git repositories, or tarballs.
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: "Parts canpoint to …"

File: templates/home/_fsf_c.html [hunk 4]
```diff
@@ -44,6 +44,6 @@ <h4>Here’s how dosbox uses it:</h4>
 
   <div class="p-flow-details__continue">
     In just a few steps, you’ll have an example C/C++ app in the Snap Store.
-    <a class="p-button--positive is-inline" href="https://docs.snapcraft.io/build-snaps/c">Continue</a>
+    <a class="p-button--positive is-inline" href="/first-snap/c">Continue</a>
```
Comment (evandandrea) [inline templates/home/_fsf_c.html]: This should go in a fsf-c branch.

File: templates/home/_fsf_go.html [hunk 5]
```diff
@@ -30,6 +30,8 @@ <h4>Here’s how httplab uses it:</h4>
     <b>go-importpath</b>: github.com/gchaincl/httplab
     <b>source</b>: .
     <b>source-type</b>: git
+	  <b> build-packages</b>:
+	    - gcc
```
Comment (evandandrea) [inline templates/home/_fsf_go.html]: This should go in the fsf-go branch.

File: webapp/first_snap/content/java/package.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        The gradle plugin can build the application using standard parameters. In this case however we have overridden some gradle parameters (to disable testing, and prevent a new git tag being created), and set an appropriate JAVA_HOME in a build: script snippet.
+    - text: |
+        In the install: script snippet we’re unpacking the built application into a directory which later gets incorporated into the final snap, defined by the $SNAPCRAFT_PART_INSTALL variable.
+    - text: |    
+        The build requires an appropriate JDK/JRE and the install step requires the addition of the unzip command, so these are specified as build-packages.
+  apps:
+    - text: |
+        Apps are the commands and services exposed to end users. If your command name matches the snap name, users will be able run the command directly.
+    - text: |
+        If you don’t want your command prefixed you can request an alias for it on the Snapcraft forum. These are set up automatically when your snap is installed from the Snap Store.
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: I would drop this point. It's too much information in this introduction to snapping. We also use "prefixed" to mean something else in the paragraph following.

File: webapp/first_snap/content/java/package.yaml (hunk 3 above)
Comment (evandandrea) [reply webapp/first_snap/content/java/package.yaml]: Still an outstanding issue.

File: webapp/first_snap/content/java/package.yaml (hunk 1 above)
Comment (evandandrea) [reply webapp/first_snap/content/java/package.yaml]: Still an issue.

File: webapp/first_snap/content/java/package.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+download: git clone https://github.com/snapcraft-docs/freeplane
+createDir: |
+  cd freeplane
+  $EDITOR snapcraft.yaml
+metadata: |
+  name: freeplane
+  version: git
+  summary: Application for Mind Mapping, Knowledge and Project Management
+  description: |
+    Freeplane is a free and open source software application that supports thinking, sharing information and getting things done at work, in school and at home. The core of the software is tools for mind mapping (also known as concept mapping or information mapping) and using mapped information.
```
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/java/package.yaml]: Still an issue.

File: templates/home/_fsf_java.html [hunk 8]
```diff
@@ -5,10 +5,40 @@ <h4>Why are snaps good for Java projects?</h4>
       <li>Simplify installation instructions, regardless of distribution, to snap install myjavaapp.</li>
       <li>Directly control the delivery of automatic application updates.</li>
     </ul>
+	<p>Distributing a Java application for Linux and reaching the widest possible audience was complicated. Typically, the user had to make sure they had the JRE/SDK version and their environment configured correctly. When a Linux distribution changes the delivered JRE, this can be problematic for applications. Snapcraft ensures the correct JRE is shipped alongside the application at all times.</p>
+  </div>
+
+  <div class='col-6'>
+    <h4>How freeplane defines snapcraft.yaml:</h4>
```
[the remark this answers is not in the document]
Comment (evandandrea) [reply templates/home/_fsf_java.html]: I cannot check whether this is still an issue due to the missing _fsf_c and _fsf_go.

File: templates/home/_fsf_java.html [hunk 9]
```diff
@@ -5,10 +5,40 @@ <h4>Why are snaps good for Java projects?</h4>
       <li>Simplify installation instructions, regardless of distribution, to snap install myjavaapp.</li>
       <li>Directly control the delivery of automatic application updates.</li>
     </ul>
+	<p>Distributing a Java application for Linux and reaching the widest possible audience was complicated. Typically, the user had to make sure they had the JRE/SDK version and their environment configured correctly. When a Linux distribution changes the delivered JRE, this can be problematic for applications. Snapcraft ensures the correct JRE is shipped alongside the application at all times.</p>
```
[the remark this answers is not in the document]
Comment (evandandrea) [reply templates/home/_fsf_java.html]: I cannot check whether this is still an issue due to the missing _fsf_c and _fsf_go.

[the remark this answers is not in the document]
Comment (evandandrea) [reply templates/home/_fsf_java.html]: This past tense is still an issue.

File: templates/home/_fsf_java.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply templates/home/_fsf_java.html]: This is still an issue.

File: webapp/first_snap/content/java/package.yaml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: This should be wrapped, but it's not:
screenshot 2019-02-11 at 11 48 23

File: webapp/first_snap/content/java/package.yaml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        The gradle plugin can build the application using standard parameters. In this case however we have overridden some gradle parameters (to disable testing, and prevent a new git tag being created), and set an appropriate JAVA_HOME in a build: script snippet.
+    - text: |
+        In the install: script snippet we’re unpacking the built application into a directory which later gets incorporated into the final snap, defined by the $SNAPCRAFT_PART_INSTALL variable.
+    - text: |    
+        The build requires an appropriate JDK/JRE and the install step requires the addition of the unzip command, so these are specified as build-packages.
+  apps:
+    - text: |
+        Apps are the commands and services exposed to end users. If your command name matches the snap name, users will be able run the command directly.
+    - text: |
+        If you don’t want your command prefixed you can request an alias for it on the Snapcraft forum. These are set up automatically when your snap is installed from the Snap Store.
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: I would drop this entire line about prefixing apps. It's too much information for an introduction to snapping.

File: webapp/first_snap/content/java/package.yaml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    - text: |
+        We have prefixed the FreePlane launch script with the desktop-launch helper script, provided by the desktop-glib-only remote part.
+    - code: |
+          apps:
+            freeplane:
+                command: desktop-launch $SNAP/freeplane-1.6.10/freeplane.sh
+    - warning: |
+        <span class="p-notification__status">Snap names are globally unique.</span> You should change any instance of <code>freeplane</code> to <code>freeplane-{name}</code> (where <code>{name}</code> is your name) before continuing to avoid conflicting with the official freeplane snap.
+    - text: The <code>command</code> specifies the path to the binary to be run. This is resolved relative to the root of your snap contents and automatically searches in the usr/sbin, usr/bin, sbin, and bin subdirectories of your snap.
+    - text: If your command name matches the snap name, users will be able run the command directly.
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: You're repeating a point you made a few lines up (matching names).

File: webapp/first_snap/content/java/package.yaml (hunk 10 above)
Comment (igorljubuncic) [reply webapp/first_snap/content/java/package.yaml]: I've uploaded a new revision. Will fix the wrapping.

File: templates/home/_fsf_java.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply templates/home/_fsf_java.html]: Closer, but needs some more tweaking:

screenshot 2019-02-11 at 15 10 23

File: webapp/first_snap/content/java/package.yaml [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+download: git clone https://github.com/snapcraft-docs/freeplane
+createDir: |
+  cd freeplane
+  $EDITOR snapcraft.yaml
+metadata: |
+  name: freeplane
+  version: '1.6.10'
+  summary: Application for Mind Mapping, Knowledge and Project Management
+  description: |
+    Freeplane is a free and open source software application that supports thinking, sharing information and getting things done
```
Comment (evandandrea) [inline webapp/first_snap/content/java/package.yaml]: Needs tweaking:

screenshot 2019-02-11 at 15 11 48

**Comment by igorljubuncic** [comment]
Made changes as requested.

File: webapp/first_snap/content/java/package.yaml (hunk 10 above)
Comment (evandandrea) [reply webapp/first_snap/content/java/package.yaml]: That commit is missing. It's still not showing correctly. Make sure you `git add`ed the files to stage, committed, and pushed.

**Comment by igorljubuncic** [comment]
Pushed the latest revision with requested changes.

File: templates/home/_fsf_java.html [hunk 14]
```diff
@@ -5,10 +5,50 @@ <h4>Why are snaps good for Java projects?</h4>
...
+      unzip DIST/freeplane_bin-*.zip -d \
+	  $SNAPCRAFT_PART_INSTALL/
+    <b>build-packages:</b>
+      - unzip
+      - openjdk-8-jdk
+
+<b>apps</b>:
+  <b>freeplane</b>:
+    <b>command</b>: desktop-launch $SNAP/freeplane-1.6.10/ \
+	freeplane.sh</pre></code>
```
Comment (evandandrea) [inline templates/home/_fsf_java.html]: screenshot 2019-02-12 at 13 02 57

As suggested above:
```
command: |
  desktop-launch \
  $SNAP/freeplane-1.6.10/freeplane.sh
```

File: webapp/first_snap/content/java/package.yaml (hunk 10 above)
Comment (evandandrea) [reply webapp/first_snap/content/java/package.yaml]: Heads up this is still not fixed in the most recent set of commits.

**Comment by igorljubuncic** [comment]
Added changes.

**Pull request merged**